docs/contributing/incremental_build.md

```diff
@@ -0,0 +1,135 @@
+# Incremental Compilation Workflow for vLLM Development
+
+When working on vLLM's C++/CUDA kernels located in the `csrc/` directory, recompiling the entire project with `pip install -e .` for every change can be time-consuming. An incremental compilation workflow using CMake allows for faster iteration by only recompiling the necessary components after an initial setup. This guide details how to set up and use such a workflow, which complements your editable Python installation.
+
+## Prerequisites
+
+Before setting up the incremental build:
+
+1. **vLLM Editable Install:** Ensure you have vLLM installed from source in an editable mode. Using pre-compiled wheels for the initial editable setup can be faster, as the CMake workflow will handle subsequent kernel recompilations.
+
+   ```bash
+   VLLM_USE_PRECOMPILED=1 uv pip install -U -e . --torch-backend=auto
+   ```
+
+2. **CUDA Toolkit:** Verify that the NVIDIA CUDA Toolkit is correctly installed and `nvcc` is accessible in your `PATH`. CMake relies on `nvcc` to compile CUDA code. You can typically find `nvcc` in `$CUDA_HOME/bin/nvcc` or by running `which nvcc`. If you encounter issues, refer to the [official CUDA Toolkit installation guides](https://developer.nvidia.com/cuda-toolkit-archive) and vLLM's main [GPU installation documentation](../getting_started/installation/gpu/cuda.inc.md#troubleshooting) for troubleshooting. The `CMAKE_CUDA_COMPILER` variable in your `CMakeUserPresets.json` should also point to your `nvcc` binary.
+
+3. **Build Tools:** Ensure the dependencies for building are installed and available, like `cmake` and `ninja`. These are installable through the `requirements/build.txt`, or can be installed by your package manager.
+
+    ```bash
+    uv pip install -r requirements/build.txt --torch-backend=auto
```

Comment:
maybe not directly related, but curious, wondering when do we recommend uv pip, when we use pip directly in the doc?

Reply:
yeah I think it's not homogeneous yet, but we're trying to replace every instance with uv. Some accelerators (tpu/gaudi/aws) though are still using pip. TPU side for instance is due to how the wheels are published, which have uv fail due to difference in package resolution.